scripts/ci/lint-clang.sh

```diff
@@ -0,0 +1,64 @@
+#!/usr/bin/env bash
+# shellcheck disable=SC1091
+#
+# Reformats C/C++ code belonging to the repository.
+# The primary Linter with support for reformatting
+# files is clang-format. Any reformatted code causes
+# non-zero exit code (CI failure).
+# Additionally, there can be secondary Linters
+# used to raise warnings.
+
+REPOROOT=$(realpath "$(dirname "$0")/../..")
+PATTERN="./*.h ./*.cpp ./*.c"
+VERSION=16
+for V in $(seq $((VERSION+10)) -1 ${VERSION}); do
```

Comment:
Well, if something fails in CI due to clang-format, it prints the version used on the server's side. Then everyone can match that major version locally. In fact, clang-format is rather stable given we run on default settings aka "LLVM style". Idea was to make it self-maintained picking up newer clang-format as we go.

Reply:
We don't want to automatically migrate to newer clang-format because that'd mean a random PR would have a massive unrelated change to merge with it. We don't care about "new features" in clang-format, just to have a consistent formatting and that's more than enough to keep on the same version for ever.

> clang-format is rather stable given we run on default settings aka "LLVM style"

Not true, unfortunately. The exact same config file can (and often does) yield different results across different versions. This is a major pain point of `clang-format` and has been discussed extensively in the list. The main developers will never make that stable on purpose, and the only alternative is to fork `clang-format`, which not many people care enough to do.

Reply:
ACK, fair enough. I will bring back logic to match a specific version.

Reply:
> a random PR would have a massive unrelated change to merge with it.

I still we are overshooting, and nobody can commit randomly formatted code because *our* installation/server determines what's correct or accepted. If we run clang-format v16, that's what counts.

Reply:
Every time we update clang-format we'll have to reformat the whole code. It makes no sense. There are no new features we need from it, so we just park on whatever version, without matching it to clang's version.

Reply:
Indeed we should only format files which are part of the incoming diff. Anyway, I will fix the version of clang-format meanwhile. Also, there are some issues introduced by this patch requiring further work.